panther_analysis_tool/log_schemas/user_defined.py

```diff
@@ -0,0 +1,279 @@
+from dataclasses import dataclass
+import fnmatch
+import json
+import logging
+import os
+from typing import List, Dict, Tuple, Optional, Any, cast
+
+import boto3
+from botocore import client
+from ruamel.yaml import YAML
+from ruamel.yaml.scanner import ScannerError
+from ruamel.yaml.parser import ParserError
+
+
+logger = logging.getLogger(__file__)
+
+
+class Client:
+    _LAMBDA_NAME = 'panther-logtypes-api'
+    _LIST_SCHEMAS_ENDPOINT = 'ListSchemas'
+    _PUT_SCHEMA_ENDPOINT = 'PutUserSchema'
+
+    def __init__(self) -> None:
+        self._lambda_client = None
+
+    @property
+    def lambda_client(self) -> client.BaseClient:
+        if self._lambda_client is None:
+            self._lambda_client = boto3.client("lambda")
+        return self._lambda_client
+
+    def list_schemas(self) -> Tuple[bool, dict]:
+        return self._invoke(
+            self._create_lambda_request(
+                endpoint=self._LIST_SCHEMAS_ENDPOINT,
+                payload={
+                    'isManaged': False
+                }
+            )
+        )
+
+    def put_schema(self, name: str, definition: str, revision: int,  # pylint: disable=too-many-arguments
+                   description: str, reference_url: str) -> Tuple[bool, dict]:
+        """
+        Update a custom schema
+        """
+        return self._invoke(
+            self._create_lambda_request(
+                endpoint=self._PUT_SCHEMA_ENDPOINT,
+                payload=dict(
+                    name=name,
+                    referenceURL=reference_url,
+                    description=description,
+                    spec=definition,
+                    revision=revision,
+                )
+            )
+        )
+
+    def _invoke(self, request: dict) -> Tuple[bool, dict]:
+        response = self.lambda_client.invoke(**request)
+        response = json.loads(response["Payload"].read().decode("utf-8"))
+        api_error = response.get("error")
+        if api_error is not None:
+            return False, response
+        return True, response
+
+    def _create_lambda_request(self, endpoint: str, payload: dict) -> dict:
+        return dict(
+            FunctionName=self._LAMBDA_NAME,
+            InvocationType="RequestResponse",
+            Payload=json.dumps({endpoint: payload})
+        )
+
+
+@dataclass
+class UploaderResult:
+    filename: str
+    name: Optional[str]
+    api_response: Optional[Dict[str, Any]] = None
+    definition: Optional[Dict[str, Any]] = None
+    error: Optional[str] = None
+    existed: Optional[bool] = None
```

Comment:
It would be useful to add some small comments what each parameter contains, for future reference.